If included file is a symlink, set basedir in utils.path_dwim_relative() to the real path

iljabauer opened the pull request:
If a role has following code in the main.yml:

```
- include: test.yml
```

and test.yml is a symlink and has following content:

```
- template: src=test.j2 dest=tmp
```

The file will not be found because the file will be searched in the folder of the main.yml.

With this change the file will be searched in the real folder of the file - like this is done with directories in https://github.com/ansible/ansible/pull/4098

iljabauer committed:
Check if task file is a symlink, if so set basedir

iljabauer committed:
Improved solution to first search the normal way and then search in symlinked path. This allow simple template inheritance

amenonsen commented:
I can confirm that this behaviour exists in current devel, but the patch no longer applies. `path_dwim_relative` has moved from `utils/__init__.py` to `parsing/__init__.py`, but the code has changed significantly.

@iljabauer any chance you could revise this to apply to current devel?

iljabauer commented:
Yes, I will have a look in to the new code

iljabauer commented:
I looked at the new code. The fix is not so easy anymore, because path_dwim_relative doesn't get the original file path anymore. This means that inside the method, we don't know that the file is imported from a symlinked file.

This information is now stored in ActionModule._task._task_include. I can pass this information inside the path_dwim_relative as optional named parameter, but then we would have to do this in every place where we need this behaviour.

It would be easier if we pass this information to the DataLoader class, than its accessible for every module which uses path_dwim_relative.

I can do this, but we have to decide, which solution is more preferable.

jimi-c commented:
Any follow-ups on this? There were recent changes to dwim functions to make improvements, so wondering if this is still an issue.

cc @bcoca

bcoca commented:
sounds like we just need to set basedir to the original path instead of the resolved one?

iljabauer commented:
First solution for this would be to extend the 'get_search_path' method of Task class. I implemented this and it is working like expected, but the depth is limited to one. I can't update the pull request, because I deleted my fork a while ago. Should I open a new pull request?

bcoca commented:
so we seem to have fixed this at one point, it is working in current devel. So i'm closing this issue.
